File: vuln-sandbox/ruby/cwe-22/path-traversal-safe.rb
# CWE-22: Path Traversal - Safe Pattern
# Rule: ruby-path-traversal-file-read
# Expected: TrueNegative

class DocumentsController < ApplicationController
  UPLOAD_DIR = "/uploads"

  def show
    # SAFE: Validate path stays within allowed directory
    filename = File.basename(params[:filename])
    safe_path = File.join(UPLOAD_DIR, filename)
    raise "Invalid path" unless safe_path.start_with?(UPLOAD_DIR)
    content = File.read(safe_path)
    render plain: content
  end

  def download
    # SAFE: Static path, no user input
    send_file "/data/public_document.pdf"
  end
end

File: vuln-sandbox/php/cwe-22/path-traversal-safe.php
<?php
// CWE-22: Path Traversal - Safe Pattern
// Rule: php-path-traversal-file-get-contents
// Expected: TrueNegative

define('UPLOAD_DIR', '/uploads/');

// SAFE: Validate path stays within allowed directory
$filename = basename($_GET['file']);
$safe_path = realpath(UPLOAD_DIR . $filename);
if ($safe_path && strpos($safe_path, realpath(UPLOAD_DIR)) === 0) {
    $content = file_get_contents($safe_path);
    echo htmlspecialchars($content);
} else {
    http_response_code(403);
    echo "Access denied";
}

// SAFE: Static path, no user input
$log = file_get_contents('/var/log/app.log');
echo htmlspecialchars($log);
?>
